src/main/kotlin/com/haroldadmin/cnradapter/NetworkResponse.kt

```diff
@@ -1,61 +1,105 @@
+ * If either of these conditions are true, then the response will contain [Unit] as the body.
+ * Otherwise, the [NetworkResponse] will be [NetworkResponse.Error.ServerError].
+ *
+ * If the network call failed due to:
+ * - Non-2xx response from the server, the [NetworkResponse] is [NetworkResponse.Error.ServerError]
+ * - Internet connectivity problems, the [NetworkResponse] is [NetworkResponse.Error.NetworkError]
+ * - Any other problems (e.g. Serialization errors), the [NetworkResponse] is [NetworkResponse.Error.UnknownError].
+ */
+public sealed interface NetworkResponse<S, E> {
     /**
-     * A request that resulted in a response with a 2xx status code that has a body.
-     */
-    data class Success<T : Any>(
-        val body: T,
-        val headers: Headers? = null,
-        val code: Int
-    ) : NetworkResponse<T, Nothing>()
-
-    /**
-     * Describe an error without a specific type.
-     * Makes it easier to deal with the case where you just want to know that an error occurred,
-     * without knowing the type
-     *
-     * @example
-     * val response = someNetworkAction()
+     * The result of a successful network request.
      *
-     * when (response) {
-     *    is NetworkResponse.Success -> // Action Succeeded do something with body
+     * If you expect your server response to not contain a body, set the success body type ([S]) to [Unit],
+     * or ensure that the response code is 204.
      *
-     *    is NetworkResponse.Error -> // Action failed do something with error
-     * }
+     * @param body The parsed body of the successful response.
+     * @param response The un-parsed [Response] from Retrofit
      */
-    interface Error {
-        val error: Throwable
-    }
+    public class Success<S, E>(
+        public val body: S,
+        public val response: Response<*>
+    ) : NetworkResponse<S, E> {
+        /**
+         * The status code returned by the server.
+         *
+         * Alias for [Response.code] of the original response
+         */
+        public val code: Int
+            get() = response.code()
 
-    /**
-     * A request that resulted in a response with a non-2xx status code.
-     */
-    data class ServerError<U : Any>(
-        val body: U?,
-        val code: Int,
-        val headers: Headers? = null
-    ) : NetworkResponse<Nothing, U>(), Error {
-        override val error = IOException("Network server error: $code \n$body")
+        /**
+         * The headers returned by the server.
+         *
+         * Alias for [Response.headers] of the original response
+         */
+        public val headers: Headers
+            get() = response.headers()
     }
 
     /**
-     * A request that didn't result in a response.
+     * The result of a failed network request.
      */
-    data class NetworkError(override val error: IOException) :
-        NetworkResponse<Nothing, Nothing>(), Error
+    public sealed interface Error<S, E> : NetworkResponse<S, E> {
```

Comment:
Should the errors be nested? and if they are nested can they be renamed to not include the "Error" keyword?
`NetworkResponse.ServerError` or `NetworkResponse.Error.Server` instead of NetworkResponse.Error.ServerError`

Reply:
I like the idea of making the naming shorter. Would that be possible without breaking backward compatibility? Would type aliases do the trick?

Reply:
They would, but isn't that the same as what we have now?

Reply:
How will it force us? isn't the error classes the same as what we have now?

Reply:
I though you were suggesting renaming these:
- `NetworkResponse.ServerError` to `NetworkResponse.Error.Server`
- `NetworkResponse.NetworkError` to `NetworkResponse.Error.Network`
- `NetworkResponse.UnknownError` to `NetworkResponse.Error.Unknown`

In which case a statement like the following:
```
when(response) {
    is NetworkReponse.Success -> ...
    is NetworkResponse.ServerError -> ...
    is NetworkResponse.NetworkError -> ...
    is NetworkResponse.UnknownError -> 
}
```
would  have to be changed to:
```
when(response) {
    is NetworkReponse.Success -> ...
    is NetworkResponse.Error.Server -> ...
    is NetworkResponse.Error.Network -> ...
    is NetworkResponse.Error.Unknown -> 
}
```
In my case that would force me to go through my project and fix a whole bunch of issues. However, if the exhaustive Error interface is released first I will have some time to replace many of those (although not all) with:
```
when(response) {
    is NetworkReponse.Success -> ...
    is NetworkResponse.Error -> ... 
}
```
which will make renaming the classes less impactful later on.

Reply:
Why not keep it?
```
    is NetworkResponse.Error -> ...
    is NetworkResponse.ServerError -> ...
    is NetworkResponse.NetworkError -> ...
    is NetworkResponse.UnknownError -> ...
 ```

Reply:
> Should the errors be nested? and if they are nested can they be renamed to not include the "Error" keyword? `NetworkResponse.ServerError` or `NetworkResponse.Error.Server` instead of NetworkResponse.Error.ServerError`

Sorry, that was how I originally implemented it. I had not realised @haroldadmin had **nested** the classes already. I would agree with your last comment: we should neither nest nor rename the classes in this release.

Reply:
@argenkiwi @gilgoldzweig I agree, the renamed and nested classes will make it painful to migrate to the newer release. Appreciate the feedback, I'll implement the changes required to keep the same type names.